sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala

```diff
@@ -236,6 +236,59 @@ case class ShowFunctionsCommand(
   }
 }
 
+
+/**
+ * A command for users to refresh the persistent function.
+ * The syntax of using this command in SQL is:
+ * {{{
+ *    REFRESH FUNCTION functionName
+ * }}}
+ */
+case class RefreshFunctionCommand(
+    databaseName: Option[String],
+    functionName: String)
+  extends RunnableCommand {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val catalog = sparkSession.sessionState.catalog
+    if (FunctionRegistry.builtin.functionExists(FunctionIdentifier(functionName))) {
+      throw new AnalysisException(s"Cannot refresh builtin function $functionName")
+    }
+    if (catalog.isTemporaryFunction(FunctionIdentifier(functionName, databaseName))) {
+      throw new AnalysisException(s"Cannot refresh temporary function $functionName")
+    }
+
+    // we only refresh the permanent function.
+    // there are 4 cases:
+    // 1. registry exists externalCatalog exists
+    // 2. registry exists externalCatalog not exists
+    // 3. registry not exists externalCatalog exists
+    // 4. registry not exists externalCatalog not exists
+    val identifier = FunctionIdentifier(
+      functionName, Some(databaseName.getOrElse(catalog.getCurrentDatabase)))
+    val isRegisteredFunction = catalog.isRegisteredFunction(identifier)
+    val isPersistentFunction = catalog.isPersistentFunction(identifier)
+    if (isRegisteredFunction && isPersistentFunction) {
+      // re-register function
+      catalog.unregisterFunction(identifier)
+      val func = catalog.getFunctionMetadata(identifier)
+      catalog.registerFunction(func, true)
+    } else if (isRegisteredFunction && !isPersistentFunction) {
+      // unregister function and throw NoSuchFunctionException
+      catalog.unregisterFunction(identifier)
+      throw new NoSuchFunctionException(identifier.database.get, functionName)
```

Comment:
It's a bit weird that you unregister then throw an exception. Maybe we should throw an exception without unregistering and say this is not in valid state?

Reply:
unregister is for clean the cache since the function is not exists. This is the core behavior for `refresh function`.